ui/eureka/eureka.js

```diff
@@ -2311,27 +2039,80 @@ class EurekaTracker {
         continue;
       }
       if (log.indexOf(' 03:') >= 0 || log.indexOf('00:0839:') >= 0) {
-        for (let i = 0; i < this.nmKeys.length; ++i) {
-          let nm = this.nms[this.nmKeys[i]];
-          if (log.match(nm.addRegex)) {
-            this.OnPopNM(nm);
-            continue;
-          }
-        }
-
         match = log.match(this.fairy.regex);
         if (match)
           this.AddFairy(match[1], match[2], match[3]);
       }
-      if (log.indexOf(' 04:') >= 0) {
-        for (let i = 0; i < this.nmKeys.length; ++i) {
-          let nm = this.nms[this.nmKeys[i]];
-          if (log.match(nm.removeRegex)) {
-            this.OnKillNM(nm);
-            continue;
+    }
+  }
+
+  OnFate(e) {
+    // Upon entering Eureka we usually receive the fate info before
+    // this.zoneInfo is loaded, so lets store the events until we're
+    // able to process them.
+    if (!this.zoneInfo) {
+      this.fateQueue.enqueue(e);
+      return;
+    }
+
+    switch (e.detail.eventType) {
+    case 'add':
+      console.log('add: ' + e.detail.fateID);
+      for (let i = 0; i < this.nmKeys.length; ++i) {
+        let nm = this.nms[this.nmKeys[i]];
+        if (e.detail.fateID == nm.fateID) {
+          this.OnFatePop(nm);
+          return;
+        }
+      }
+      if (this.options.TrackBunnyFates) {
+        for (let i = 0; i < this.bunnyKeys.length; ++i) {
+          let bunny = this.bunnys[this.bunnyKeys[i]];
+          if (e.detail.fateID == bunny.fateID) {
+            this.OnBunnyPop(bunny);
+            return;
+          }
+        }
+      }
+      break;
+    case 'remove':
+      console.log('remove: ' + e.detail.fateID + ' (' + e.detail.progress + '%)');
+      for (let i = 0; i < this.nmKeys.length; ++i) {
+        let nm = this.nms[this.nmKeys[i]];
+        if (e.detail.fateID == nm.fateID) {
+          this.OnFateKill(nm);
+          return;
+        }
+      }
+      if (this.options.TrackBunnyFates) {
+        for (let i = 0; i < this.bunnyKeys.length; ++i) {
+          let bunny = this.bunnys[this.bunnyKeys[i]];
+          if (e.detail.fateID == bunny.fateID) {
+            this.OnFateKill(bunny);
+            return;
           }
         }
       }
+      break;
+    case 'update':
+      console.log('update: ' + e.detail.fateID + ' (' + e.detail.progress + '%)');
+      for (let i = 0; i < this.nmKeys.length; ++i) {
+        let nm = this.nms[this.nmKeys[i]];
+        if (e.detail.fateID == nm.fateID) {
+          this.OnFateUpdate(nm, e.detail.progress);
+          return;
+        }
+      }
+      if (this.options.TrackBunnyFates) {
+        for (let i = 0; i < this.bunnyKeys.length; ++i) {
+          let bunny = this.bunnys[this.bunnyKeys[i]];
```

Comment:
One question I have is if the bunnies can be less special cased here.  I think removing the bunny option would help.

I think if you put the bunnies in the nm list proper (maybe with a bunny: true setting to change the sound) then this code could just be one loop.

Additionally, instead of having a bunch of code have to differentiate `.bunny-pop` from `.nm-pop`, you could add a `.bunny` class to the bunny nms, and treat bunnies as nms with an additional class.  In other words, bunny fates get `.nm-pop` and `.nm-down` but you just add another css rule for `.bunny.nm-pop` that changes the text color.  I think this would clean up some special casing code too.

And then OnFatePop could just be like "is this fate a bunny? pick a different sound" rather than having to duplicate some code in OnBunnyPop.

What do you think?

Reply:
Yeah that sounds much, much cleaner.

Reply:
Implemented as suggested.